### PR TITLE
feat: add construction order of instances to the callback identification

### PR DESCRIPTION
**Pull request by hsgwa**
Signed-off-by: hsgwa <19860128+hsgwa@users.noreply.github.com>

## Description

Currently, CARET uses the following information to identify callbacks

- node_name
- callback type: timer/ subscription
- paramter: period/ topic_name
- symbol

However, there are callbacks that cannot be uniquely identified in Autoware.

This PR adds information on the order of creation of instances only for callbacks that cannot be identified by callback, parameter, or symbol.

- node_name
- callback type: timer/ subscription
- paramter: period/ topic_name
- symbol
- construction order  [Optional] (new)

## Notes for reviewers

- This PR does not destroy backward compatibility.

We can check with old Autoware data by modifying processor.py as follows

```
    def _handle_rcl_timer_init(
        self,
        event: Dict,
    ) -> None:
        handle = get_field(event, 'timer_handle')
        timestamp = get_field(event, '_timestamp')
        period = get_field(event, 'period')
        tid = get_field(event, '_vtid')
        if period == 20000000 - 1:
            period = 20000000
        self.data.add_timer(tid, handle, timestamp, period)
```

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] (Optional) The PR has been properly tested with CARET_report verification.
- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

**Review by takeshi-iwanari — COMMENTED** [review]
@hsgwa 
- I attached two files:
  - autoware_launch_trace_20221222-181030.zip: trace log for Autoware.Universe where timer callback parameters are not modified (duplicated parameters exist). It will be useful for testing.
  - log_architecture.zip: architecture file and log file (I paste some screenshots below)
- I confirmed:
  - Warnings for `Duplicate parameter callback` disappear :+1: 
  - `construction_order` is nicely added :+1: 
image
image
- Question:
  - However, `Several candidates were found` warnings still exist :thinking:  
  - Does this PR solve subscription callbacks with the same topic ?
  - It looks symbols are different. Is this the reason why `construction_order` is not added? Another PR handles this case?
image


[log_architecture.zip](https://github.com/tier4/CARET_analyze/files/10285726/log_architecture.zip)
[autoware_launch_trace_20221222-181030.zip](https://github.com/tier4/CARET_analyze/files/10285727/autoware_launch_trace_20221222-181030.zip)

**Comment by hsgwa** [comment]
@takeshi-iwanari 
> However, Several candidates were found warnings still exist
> Does this PR solve subscription callbacks with the same topic ?

I modified to address it.
This PR solves subscription callbacks with the same topic by using `callback_name`.

> It looks symbols are different. Is this the reason why construction_order is not added? Another PR handles this case?

Yes. construction_order is added only if following attributes are same.

- node_name
- callback type: timer/ subscription
- paramter: period/ topic_name
- symbol

**Review by takeshi-iwanari — CHANGES_REQUESTED** [review]
@hsgwa 
I tested communication/path analysis using [this tracedata](https://github.com/tier4/CARET_analyze/files/10285727/autoware_launch_trace_20221222-181030.zip), where `/mission_planner` node has two subscription callback functions for `/kinematic_state` topic.
There are two problems:
- `get_communication` returns error: `ItemNotFoundError: Arguments may be wrong. Aren't they bellow?`
- `search_paths` returns error: `MultipleItemFoundError: Failed to identify item.`

```py
src_node = '/localization/pose_twist_fusion_filter/stop_filter'
dst_node = '/planning/mission_planning/mission_planner'    # ERROR
# dst_node = '/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner'    # OK
topic_name = '/localization/kinematic_state'

comm = app.get_communication(src_node, dst_node, topic_name)
target_path_list = arch.search_paths(src_node, dst_node, max_node_depth=5)
```

--- Added ---

- Result of `app.get_communications(topic_name)` doesn't include communication to `/mission_planner`
- Result of `app.get_subscriptions(topic_name)` doesn't include `/mission_planner`

File: src/caret_analyze/architecture/architecture_exporter.py [hunk 1]
```diff
@@ -109,24 +109,31 @@ def _timer_cb_to_dict(
...
     ) -> Dict:
-        return {
+        d: Dict[str, Union[str, int]]
+        d = {
             'callback_name': subscription_callback.callback_name,
             'callback_type': str(CallbackType.SUBSCRIPTION),
             'topic_name': subscription_callback.subscribe_topic_name,
             'symbol': subscription_callback.symbol,
         }
+        if subscription_callback.construction_order > 0:
```
Comment (takeshi-iwanari) [inline src/caret_analyze/architecture/architecture_exporter.py:133]:
```suggestion
        if subscription_callback.construction_order and subscription_callback.construction_order > 0:
```

**Review by takeshi-iwanari — CHANGES_REQUESTED** [review]
I encountered error when `arch.export` after `arch.add_path`

```py
target_path_list = arch.search_paths(*target_path_info, max_node_depth=5)
arch.add_path(target_path_name, target_path_list[0])
arch.export(architecture_path, force=True)
```

```
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_2086911/1057420871.py in <module>
      3 target_path_list = arch.search_paths(*target_path_info, max_node_depth=5)
      4 arch.add_path(target_path_name, target_path_list[0])
----> 5 arch.export(architecture_path, force=True)

~/ros2_caret_ws/build/caret_analyze/caret_analyze/architecture/architecture.py in export(self, file_path, force)
    229         exporter = ArchitectureExporter(
    230             self.nodes, self.executors, self.paths, force)
--> 231         exporter.execute(file_path)
    232 
    233     def search_paths(

~/ros2_caret_ws/build/caret_analyze/caret_analyze/architecture/architecture_exporter.py in execute(self, file_path)
     47         mode = 'w' if self._force else 'x'
     48         with open(file_path, mode=mode) as f:
---> 49             f.write(str(self))
     50 
     51     def __str__(self) -> str:

~/ros2_caret_ws/build/caret_analyze/caret_analyze/architecture/architecture_exporter.py in __str__(self)
     51     def __str__(self) -> str:
     52         import yaml
---> 53         obj = self.to_dict()
...
--> 133         if subscription_callback.construction_order > 0:
    134             d['construction_order'] = subscription_callback.construction_order
    135 

TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

File: src/caret_analyze/architecture/architecture_exporter.py (hunk 1 above)
Comment (hsgwa) [reply src/caret_analyze/architecture/architecture_exporter.py:133]: fixed in :[8ceaf80](https://github.com/tier4/CARET_analyze/pull/225/commits/8ceaf80c437b7b8155618eeae5e168b34c10bbc7)

File: src/caret_analyze/architecture/architecture_exporter.py [hunk 2]
```diff
@@ -109,24 +109,31 @@ def _timer_cb_to_dict(
...
     ) -> Dict:
-        return  \
-            {
+        d = {
                 'callback_name': timer_callback.callback_name,
                 'callback_type': str(CallbackType.TIMER),
                 'period_ns': timer_callback.period_ns,
                 'symbol': timer_callback.symbol,
             }
+        if timer_callback.construction_order > 0:
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply src/caret_analyze/architecture/architecture_exporter.py:118]: fixed in :[8ceaf80](https://github.com/tier4/CARET_analyze/pull/225/commits/8ceaf80c437b7b8155618eeae5e168b34c10bbc7)

**Comment by hsgwa** [comment]
@takeshi-iwanari 
As you indicated, this PR is unable to resolve get_communication and get_subscription.
We are considering addressing this issue in a future PR.
In the meantime, we have fixed the architecture.export error.

**Review by takeshi-iwanari — APPROVED** [review]
I confirmed that the PR can analyze a node:
  - which has several subscription callback functions with the same topic name
  - which has several timer callback functions with the same timer period time

**Review by keita1523 — COMMENTED** [review]
@hsgwa 
Please confirm my review.

File: src/caret_analyze/infra/lttng/lttng_info.py [hunk 3]
```diff
@@ -961,36 +965,42 @@ def _build_timer_callbacks(
...
         ]
 
         def callback_id(row: pd.Series) -> str:
             cb_object = row['callback_object']
             return f'timer_callback_{cb_object}'
         timers = data.timers.clone()
         timers.reset_index()
+        timers.rename_column('period', 'period_ns')
 
+        merge_drop_columns = ['tid', 'rmw_handle']
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply src/caret_analyze/infra/lttng/lttng_info.py]: make sense.
addressed in [2edeba2](https://github.com/tier4/CARET_analyze/pull/225/commits/2edeba2d2f141bf889d9a0842cc8844b802a4f90)

File: src/caret_analyze/architecture/graph_search.py [hunk 4]
```diff
@@ -557,6 +562,8 @@ def __init__(
 
             self._graph.add_node(GraphNode(node.node_name))
 
+        duplicated_comms = {}
```
[the remark this answers is not in the document]
Comment (hsgwa) [reply src/caret_analyze/architecture/graph_search.py]: addressed in [2edeba2](https://github.com/tier4/CARET_analyze/pull/225/commits/2edeba2d2f141bf889d9a0842cc8844b802a4f90)

**Review by keita1523 — APPROVED** [review]
LGTM!

**Pull request merged**